dsdl_compiler/libcanard_dsdl_compiler/__init__.py

```diff
@@ -305,46 +318,42 @@ def inject_cpp_types(attributes):
             if (count == length):
                 a.last_item = True
 
-            data = type_to_c_type(a.type)
+            data = type_to_c_type(a.data_type)
             for key, value in data.items():
                 setattr(a, key, value)
 
-            if a.type.category == t.CATEGORY_ARRAY:
-                a.array_size = a.type.max_size
+            if isinstance(a.data_type, ArrayType):
+                a.array_size = a.max_array_elements
                 has_array = True
 
-            a.type_category = a.type.category
-            a.void = a.type.category == a.type.CATEGORY_VOID
-            if a.void:
-                assert not a.name
-                a.name = ''
+            a.void = isinstance(a.data_type, VoidType)
         return has_array
 
     def has_float16(attributes):
         has_float16 = False
         for a in attributes:
-            if a.type.category == t.CATEGORY_PRIMITIVE and a.type.kind == a.type.KIND_FLOAT and a.bitlen == 16:
+            if isinstance(a.data_type, FloatType) and a.bitlen == 16: # TODO: Fixme
                 has_float16 = True
         return has_float16
 
-    if t.kind == t.KIND_MESSAGE:
+    if isinstance(t, (StructureType, UnionType)):
         t.has_array = inject_cpp_types(t.fields)
         t.has_float16 = has_float16(t.fields)
         inject_cpp_types(t.constants)
         t.all_attributes = t.fields + t.constants
-        t.union = t.union and len(t.fields)
+        t.union = isinstance(t, UnionType)
         if t.union:
             t.union = len(t.fields).bit_length()
```

Comment:
Will break if `len(t.fields)` is a power of two. The correct expression is `(len(t.fields) - 1).bit_length()`. Per the spec, a union cannot contain less than 2 fields, this is enforced by PyDSDL.